src/main/java/org/broadinstitute/hellbender/utils/spark/SparkUtils.java

```diff
@@ -103,4 +123,358 @@ public static boolean pathExists(final JavaSparkContext ctx, final Path targetPa
+        // subsequent partitions are coalesced if needed, so for each partition p find the latest subsequent
+        // partition that is needed to read all of the intervals that start in p.
+        List<Integer> maxEndPartitionIndexes = new ArrayList<>();
+        for (int i = 0; i < reads.getNumPartitions(); i++) {
+            maxEndPartitionIndexes.add(i);
+        }
+        OverlapDetector<PartitionLocatable<SimpleInterval>> overlapDetector = OverlapDetector.create(partitionReadExtents);
+        List<PartitionLocatable<I>> indexedIntervals = new ArrayList<>();
+        for (I interval : intervals) {
+            int[] partitionIndexes = overlapDetector.getOverlaps(interval).stream()
+                    .mapToInt(PartitionLocatable::getPartitionIndex).toArray();
+            if (partitionIndexes.length == 0) {
+                // interval does not overlap any partition - skip it
+                continue;
+            }
+            Arrays.sort(partitionIndexes);
+            int startIndex = partitionIndexes[0];
+            int endIndex = partitionIndexes[partitionIndexes.length - 1];
+            indexedIntervals.add(new PartitionLocatable<I>(startIndex, interval));
+            if (endIndex > maxEndPartitionIndexes.get(startIndex)) {
+                maxEndPartitionIndexes.set(startIndex, endIndex);
+            }
+        }
+
+        JavaRDD<R> coalescedRdd = coalesce(reads, readClass, new RangePartitionCoalescer(maxEndPartitionIndexes));
+
+        // Create an RDD of intervals with the same number of partitions as the reads, and where each interval
+        // is in its start partition.
+        JavaRDD<I> intervalsRdd = ctx.parallelize(indexedIntervals)
+                .mapToPair(interval ->
+                        new Tuple2<>(interval.getPartitionIndex(), interval.getLocatable()))
+                .partitionBy(new KeyPartitioner(reads.getNumPartitions())).values();
+
+        // zipPartitions on coalesced read partitions and intervals, and apply the function f
+        return coalescedRdd.zipPartitions(intervalsRdd, f);
+    }
+
+    /**
+     * Turn a pair of iterators over intervals and reads, into a single iterator over pairs made up of an interval and
+     * the reads that overlap it. Intervals with no overlapping reads are dropped.
+     */
+    static <R extends Locatable, I extends Locatable> Iterator<Tuple2<I, Iterable<R>>> readsPerShard(Iterator<R> reads, Iterator<I> shards, SAMSequenceDictionary sequenceDictionary) {
+        if (!shards.hasNext()) {
+            return Collections.emptyIterator();
+        }
+        PeekingIterator<R> peekingReads = Iterators.peekingIterator(reads);
+        PeekingIterator<I> peekingShards = Iterators.peekingIterator(shards);
+        Iterator<Tuple2<I, Iterable<R>>> iterator = new AbstractIterator<Tuple2<I, Iterable<R>>>() {
+            // keep track of current and next, since reads can overlap two shards
```

Comment:
We're assuming that shards are larger than reads (and padding is too). Since shard and padding size is configurable, this shouldn't be a problem. I'll add a fail-fast check for this constraint, as we discussed.